### PR TITLE
Relocate Role-Based Configuration classes to make them accessible when creating configurations

### PR DESCRIPTION
**Pull request by tresat**
`RoleBasedConfigurationContainerInternal` needs to exist in a deeper part of the subproject hierarchy than `dependency-management`, so that the factory method in it can be called from every Gradle subproject creating configurations.

After moving this class to the `core` subproject, the class hierarchy is changed so that `RBCCI` extends `ConfigurationContainerInternal`, instead of the opposite.  This means the create factory methods must return a plain `Configuration`, rather than `ConfigurationInternal`, which is not visible from `core`.  The new internal `ConfigurationRole` interface and `ConfigurationRoles` implementation enum also follow `RBCCI` into core.

The `isCanBeDeclaredAgainst` and `setCanBeDeclaredAgainst` methods, which were previously added to the `ConfigurationInternal` interface to avoid adding to the public `Configuration` API, are now moved to `Configuration` (this puts them next to the equivalent methods for isCanBeConsumed/Resolved, which is a more appropriate place anyway) to be accessible.  They are now marked `@Incubating` since 8.1 for binary compatibility check reasons, though they existed in the internal type hierarchy prior to this.

The `preventUsageMutation` method also has to be accessible from `RBCCI` in core, so it can't remain on `ConfigurationInternal` and so it moves to `DeprecableConfiguration`.  As the `RBCCI` is the only caller of this method outside of `DefaultConfiguration` itself, this shouldn't be a problem.  Even if it makes the interface slightly less cohesive, it doesn't seem worth creating a different new, accessible, sub-interface of Configuration just for this method.

In summary, strictly internal relocations, with the exception of exposing `isCanBeDeclaredAgainst` and `setCanBeDeclaredAgainst` on `Configuration`, which is arguably where they should have been added in the first place.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
From our conversation yesterday.  I don't think the failing test here is relevant.

**Comment by jvandort** [comment]
> @big-guy, @jvandort - From our conversation yesterday. I don't think the failing test here is relevant.

Rebase onto latest master and I think that should go away

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 1]
```diff
@@ -543,6 +543,24 @@ public String determineName(Configuration c) {
      */
     boolean isCanBeResolved();
 
+    /**
+     * Configures if a configuration can have dependencies declared upon it.
+     *
+     * @since 8.1
+     */
+    @Incubating
+    void setCanBeDeclaredAgainst(boolean allowed);
```
Comment (jvandort) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java:552]: I've put some prior thought into these names. I think that `is/setCanBeDeclaredUpon` would be more fitting, since the phrase "declare a dependency on <configuration name>" is the preferred terminology as opposed to "declare a dependency against <configuration name>"

Thoughts here? Its not a big deal and I don't think we need to deal with this here (since it would imply changes to tests and other existing first-party plugin implementations). But I thought I'd bring this up since we're now exposing this as part of a public API.

**Review by jvandort — APPROVED** [review]

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java (hunk 1 above)
Comment (tresat) [reply subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java:552]: I talked with Sterling about this, and I'm going to try a couple of things to avoid adding to the public Configuration API.  I think the name will matter less then.

But I have no strong feelings about the name.  If renamin isn't a pain, I'll include it in this change.  If it is, I'll do it as part of the series of changes after this.

**Comment by tresat** [comment]
@bot-gradle test and merge
